File: src/assembler/agents-assembler.ts
/**
 * AgentsAssembler — assembles .claude/agents/ from templates based on project config.
 *
 * Migrated from Python `assembler/agents.py` (263 lines).
 * Selection logic lives in {@link ./agents-selection.ts}; this module handles file I/O.
 *
 * @module
 */
import * as fs from "node:fs";
import * as path from "node:path";
import type { ProjectConfig } from "../models.js";
import { TemplateEngine } from "../template-engine.js";
import {
  copyTemplateFile,
  copyTemplateFileIfExists,
} from "./copy-helpers.js";
import type { AssembleResult } from "./rules-assembler.js";
import {
  selectConditionalAgents,
  buildChecklistRules,
  checklistMarker,
} from "./agents-selection.js";

const AGENTS_TEMPLATES_DIR = "agents-templates";
const CORE_DIR = "core";
const CONDITIONAL_DIR = "conditional";
const DEVELOPERS_DIR = "developers";
const CHECKLISTS_DIR = "checklists";
const AGENTS_OUTPUT = "agents";
const MD_EXTENSION = ".md";

/** Assembles agent files from templates based on project config. */
export class AgentsAssembler {
  /** Scan core agents directory for .md files, sorted alphabetically. */
  selectCoreAgents(resourcesDir: string): string[] {
    const corePath = path.join(
      resourcesDir, AGENTS_TEMPLATES_DIR, CORE_DIR,
    );
    if (!fs.existsSync(corePath)) return [];
    return fs
      .readdirSync(corePath, { withFileTypes: true })
      .filter((e) => e.isFile() && e.name.endsWith(MD_EXTENSION))
      .map((e) => e.name)
      .sort();
  }

  /** Evaluate feature gates and return conditional agent filenames. */
  selectConditionalAgents(config: ProjectConfig): string[] {
    return selectConditionalAgents(config);
  }

  /** Return developer agent filename for the project language. */
  selectDeveloperAgent(config: ProjectConfig): string {
    const safeName = path.basename(config.language.name);
    return `${safeName}-developer${MD_EXTENSION}`;
  }

  /** Main entry point: assemble all agents. */
  assemble(
    config: ProjectConfig,
    outputDir: string,
    resourcesDir: string,
    engine: TemplateEngine,
  ): AssembleResult {
    const files: string[] = [];
    files.push(
      ...this.assembleCore(resourcesDir, outputDir, engine),
    );
    files.push(
      ...this.assembleConditional(config, resourcesDir, outputDir, engine),
    );
    const dev = this.copyDeveloperAgent(
      config, resourcesDir, outputDir, engine,
    );
    if (dev !== null) {
      files.push(dev);
    }
    this.injectChecklists(config, outputDir, resourcesDir);
    return { files, warnings: [] };
  }

  private copyCoreAgent(
    agentFile: string, resourcesDir: string,
    outputDir: string, engine: TemplateEngine,
  ): string {
    const src = path.join(
      resourcesDir, AGENTS_TEMPLATES_DIR, CORE_DIR, agentFile,
    );
    const dest = path.join(outputDir, AGENTS_OUTPUT, agentFile);
    return copyTemplateFile(src, dest, engine);
  }

  private copyConditionalAgent(
    agentFile: string, resourcesDir: string,
    outputDir: string, engine: TemplateEngine,
  ): string | null {
    const src = path.join(
      resourcesDir, AGENTS_TEMPLATES_DIR, CONDITIONAL_DIR, agentFile,
    );
    const dest = path.join(outputDir, AGENTS_OUTPUT, agentFile);
    return copyTemplateFileIfExists(src, dest, engine);
  }

  private copyDeveloperAgent(
    config: ProjectConfig, resourcesDir: string,
    outputDir: string, engine: TemplateEngine,
  ): string | null {
    const agentFile = this.selectDeveloperAgent(config);
    const src = path.join(
      resourcesDir, AGENTS_TEMPLATES_DIR, DEVELOPERS_DIR, agentFile,
    );
    const dest = path.join(outputDir, AGENTS_OUTPUT, agentFile);
    return copyTemplateFileIfExists(src, dest, engine);
  }

  private injectChecklists(
    config: ProjectConfig, outputDir: string, resourcesDir: string,
  ): void {
    for (const rule of buildChecklistRules(config)) {
      if (!rule.active) continue;
      this.injectSingleChecklist(
        rule.agent, rule.checklist, outputDir, resourcesDir,
      );
    }
  }

  private injectSingleChecklist(
    agentFile: string, checklistFile: string,
    outputDir: string, resourcesDir: string,
  ): void {
    const agentPath = path.join(outputDir, AGENTS_OUTPUT, agentFile);
    if (!fs.existsSync(agentPath)) return;
    const checklistSrc = path.join(
      resourcesDir, AGENTS_TEMPLATES_DIR, CHECKLISTS_DIR, checklistFile,
    );
    if (!fs.existsSync(checklistSrc)) return;
    const marker = checklistMarker(checklistFile);
    const section = fs.readFileSync(checklistSrc, "utf-8");
    const base = fs.readFileSync(agentPath, "utf-8");
    const result = TemplateEngine.injectSection(base, section, marker);
    fs.writeFileSync(agentPath, result, "utf-8");
  }

  private assembleCore(
    resourcesDir: string, outputDir: string, engine: TemplateEngine,
  ): string[] {
    return this.selectCoreAgents(resourcesDir).map(
      (agent) => this.copyCoreAgent(agent, resourcesDir, outputDir, engine),
    );
  }

  private assembleConditional(
    config: ProjectConfig, resourcesDir: string,
    outputDir: string, engine: TemplateEngine,
  ): string[] {
    const results: string[] = [];
    for (const agent of this.selectConditionalAgents(config)) {
      const copied = this.copyConditionalAgent(
        agent, resourcesDir, outputDir, engine,
      );
      if (copied !== null) results.push(copied);
    }
    return results;
  }
}

File: java/src/main/java/dev/iadev/assembler/AgentsAssembler.java
package dev.iadev.assembler;

import dev.iadev.config.ContextBuilder;
import dev.iadev.domain.model.ProjectConfig;
import dev.iadev.template.TemplateEngine;

import java.io.IOException;
import java.io.UncheckedIOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.stream.Stream;

/**
 * Assembles {@code .claude/agents/} from templates based on
 * project configuration.
 *
 * <p>Checklist injection is delegated to
 * {@link ChecklistInjector}.</p>
 *
 * @see Assembler
 * @see AgentsSelection
 * @see ChecklistInjector
 */
public final class AgentsAssembler implements Assembler {

    private static final String AGENTS_TEMPLATES_DIR =
            "agents-templates";
    private static final String CORE_DIR = "core";
    private static final String CONDITIONAL_DIR =
            "conditional";
    private static final String DEVELOPERS_DIR =
            "developers";
    private static final String AGENTS_OUTPUT = "agents";
    private static final String MD_EXTENSION = ".md";

    private final Path resourcesDir;

    /**
     * Creates an AgentsAssembler using classpath resources.
     */
    public AgentsAssembler() {
        this(resolveClasspathResources());
    }

    /**
     * Creates an AgentsAssembler with an explicit resources
     * directory.
     *
     * @param resourcesDir the base resources directory
     */
    public AgentsAssembler(Path resourcesDir) {
        this.resourcesDir = resourcesDir;
    }

    /** {@inheritDoc} */
    @Override
    public List<String> assemble(
            ProjectConfig config,
            TemplateEngine engine,
            Path outputDir) {
        Map<String, Object> context =
                ContextBuilder.buildContext(config);
        List<String> files = new ArrayList<>();

        files.addAll(assembleCore(
                outputDir, engine, context));
        files.addAll(assembleConditional(
                config, outputDir, engine, context));

        copyDeveloperAgent(
                config, outputDir, engine, context)
                .ifPresent(files::add);

        ChecklistInjector.injectChecklists(
                config, resourcesDir, outputDir);
        return files;
    }

    /**
     * Scans core agents directory for .md files, sorted
     * alphabetically.
     *
     * @return sorted list of core agent filenames
     */
    List<String> selectCoreAgents() {
        Path corePath = resourcesDir.resolve(
                AGENTS_TEMPLATES_DIR + "/" + CORE_DIR);
        if (!Files.exists(corePath)
                || !Files.isDirectory(corePath)) {
            return List.of();
        }
        try (Stream<Path> stream = Files.list(corePath)) {
            return stream
                    .filter(Files::isRegularFile)
                    .map(p -> p.getFileName().toString())
                    .filter(name -> name.endsWith(
                            MD_EXTENSION))
                    .sorted()
                    .toList();
        } catch (IOException e) {
            throw new UncheckedIOException(
                    "Failed to list core agents: "
                            + corePath, e);
        }
    }

    private List<String> assembleCore(
            Path outputDir,
            TemplateEngine engine,
            Map<String, Object> context) {
        List<String> generated = new ArrayList<>();
        for (String agent : selectCoreAgents()) {
            String result = copyCoreAgent(
                    agent, outputDir, engine, context);
            generated.add(result);
        }
        return generated;
    }

    private List<String> assembleConditional(
            ProjectConfig config,
            Path outputDir,
            TemplateEngine engine,
            Map<String, Object> context) {
        List<String> generated = new ArrayList<>();
        List<String> conditional =
                AgentsSelection.selectConditionalAgents(
                        config);
        for (String agent : conditional) {
            copyConditionalAgent(
                    agent, outputDir, engine, context)
                    .ifPresent(generated::add);
        }
        return generated;
    }

    private String copyCoreAgent(
            String agentFile,
            Path outputDir,
            TemplateEngine engine,
            Map<String, Object> context) {
        Path src = resourcesDir.resolve(
                AGENTS_TEMPLATES_DIR + "/"
                        + CORE_DIR + "/" + agentFile);
        Path dest = outputDir.resolve(
                AGENTS_OUTPUT + "/" + agentFile);
        return CopyHelpers.copyTemplateFile(
                src, dest, engine, context);
    }

    private Optional<String> copyConditionalAgent(
            String agentFile,
            Path outputDir,
            TemplateEngine engine,
            Map<String, Object> context) {
        Path src = resourcesDir.resolve(
                AGENTS_TEMPLATES_DIR + "/"
                        + CONDITIONAL_DIR + "/"
                        + agentFile);
        Path dest = outputDir.resolve(
                AGENTS_OUTPUT + "/" + agentFile);
        return CopyHelpers.copyTemplateFileIfExists(
                src, dest, engine, context);
    }

    private Optional<String> copyDeveloperAgent(
            ProjectConfig config,
            Path outputDir,
            TemplateEngine engine,
            Map<String, Object> context) {
        String agentFile =
                AgentsSelection.selectDeveloperAgent(
                        config);
        Path src = resourcesDir.resolve(
                AGENTS_TEMPLATES_DIR + "/"
                        + DEVELOPERS_DIR + "/"
                        + agentFile);
        Path dest = outputDir.resolve(
                AGENTS_OUTPUT + "/" + agentFile);
        return CopyHelpers.copyTemplateFileIfExists(
                src, dest, engine, context);
    }

    private static Path resolveClasspathResources() {
        return dev.iadev.util.ResourceResolver
                .resolveResourcesRoot(AGENTS_TEMPLATES_DIR);
    }
}
